File: app/src/main/java/com/crevion/apps/cleanandroidcode/home/HomePresenter.java
package com.crevion.apps.cleanandroidcode.home;


import com.crevion.apps.cleanandroidcode.models.CityListResponse;
import com.crevion.apps.cleanandroidcode.networking.NetworkError;
import com.crevion.apps.cleanandroidcode.networking.Service;

import rx.Subscription;
import rx.subscriptions.CompositeSubscription;

/**
 * Created by yusufaw on 1/27/18.
 */

public class HomePresenter {
    private final Service service;
    private final HomeView view;
    private CompositeSubscription subscriptions;

    public HomePresenter(Service service, HomeView view) {
        this.service = service;
        this.view = view;
        this.subscriptions = new CompositeSubscription();
    }

    public void getCityList() {
        view.showWait();

        Subscription subscription = service.getCityList(new Service.GetCityListCallback() {
            @Override
            public void onSuccess(CityListResponse cityListResponse) {
                view.removeWait();
                view.geCityListSuccess(cityListResponse);
            }

            @Override
            public void onError(NetworkError networkError) {
                view.removeWait();
                view.onFailure(networkError.getAppErrorMessage());
            }
        });
        this.subscriptions.add(subscription);
    }

    public void onStop() {
        subscriptions.unsubscribe();
    }
}

File: app/src/main/java/com/crevion/apps/cleanandroidcode/ui/home/HomePresenter.kt
package com.crevion.apps.cleanandroidcode.ui.home

import com.crevion.apps.cleanandroidcode.models.CityListResponse
import com.crevion.apps.cleanandroidcode.networking.NetworkError
import com.crevion.apps.cleanandroidcode.networking.Service
import rx.subscriptions.CompositeSubscription

/**
 * Created by yusufaw on 2/9/18.
 */

class HomePresenter(private val service: Service, private val view: HomeView) {
    private val subscriptions: CompositeSubscription = CompositeSubscription()

    fun getCityList() {
        view.showWait()

        val subscription = service.getCityList(object : Service.GetCityListCallback {
            override fun onSuccess(cityListResponse: CityListResponse) {
                view.removeWait()
                view.geCityListSuccess(cityListResponse)
            }

            override fun onError(networkError: NetworkError) {
                view.removeWait()
                view.onFailure(networkError.appErrorMessage)
            }
        })
        this.subscriptions.add(subscription)
    }

    fun onStop() {
        subscriptions.unsubscribe()
    }
}